File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<vector<int>> v) {
    int s = v.size();
    for (int i = 1; i < s; i++) {
        v[i][0] += max({v[i - 1][1], v[i - 1][2], v[i - 1][3]});
        v[i][1] += max({v[i - 1][0], v[i - 1][2], v[i - 1][3]});
        v[i][2] += max({v[i - 1][0], v[i - 1][1], v[i - 1][3]});
        v[i][3] += max({v[i - 1][0], v[i - 1][1], v[i - 1][2]});
    }
    s--;
    return max({v[s][0], v[s][1], v[s][2], v[s][3]});
}

File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.java
import java.util.*;

class Solution {
    int solution(int[][] land) {
        int answer = 0;

        for (int i = 0; i < land.length - 1; i++) {
            for (int j = 0; j < 4; j++) {
                int max = Integer.MIN_VALUE;
                for (int k = 0; k < 4; k++) {
                    if (j == k) continue;
                    max = Math.max(max, land[i][k]);
                }
                
                land[i + 1][j] += max;
            }
        }
        
        for (int j = 0; j < 4; j++) {
            answer = Math.max(answer, land[land.length - 1][j]);
        }

        return answer;
    }
}